feat: Create upgrading tool (#988)
vaadin/docs: articles/upgrading/index.asciidoc

---
title: Upgrading
order: 70
---

= Upgrading Guide
:!toc:

> Instructions for upgrading to the latest Vaadin version. To run applications or components developed with Vaadin 7 or 8 inside an application written using the latest version, see <<{articles}/tools/mpr/overview#,Multiplatform Runtime>>.

Reviewer: ```suggestion
pass:[<!-- Vale Vaadin.Versions = NO -->]

Instructions for upgrading to the latest Vaadin version. To run applications or components developed with Vaadin 7 or 8 inside an application written using the latest version, see <<{articles}/tools/mpr/overview#,Multiplatform Runtime>>.

pass:[<!-- Vale Vaadin.Versions = YES -->]
```

[source,typescript, role="render-only embedded"]
----
include::{root}/frontend/demo/upgrade-tool/upgrade-tool.ts[render, persistent]
----

////
Upgrading instructions are all loaded to the page as hidden by default.
The upgrade-tool manipulates the visibility of the instructions via their classnames depending of the user's selected options.

This is the description of the classnames:

.hidden:
    All items should be given this classname for them to be hidden when the page is loaded (the upgrade tool will remove this classname conditional on user selection).

.upgrade-instructions:
    All items should be given this classname as well. This classname will enable the upgrade tool to re-hide all the elements as the user selection changes.

.instructions-all:

>     Upgrading instructions that should appear regardless of the user's selected configurations.

Reviewer: I wonder if we could make this somehow a bit more automated. Having the author write all of these repeatedly seems quite verbose and tedious. Not a big issue, copy-paste is easy, but just wondering if we can make it easier for dev teams to author these…

Reviewer: I think my approach would be something like this:

- Mark the version range only on the sections
- Annotate Flow/Fusion/etc. specific instructions with `.flow`, `.fusion`, no version range.  
Side note: the way you do it now won't work for longer pieces of content, the class will only apply on the first item in the included content. I used an openblock (`--`) to get around that.

For example:

```asciidoc
[.v14-22]
=== Upgrade Steps | 14 → 22

[.flow]
--
include::flow-instructions/_upgrade-14-22.asciidoc[]
--

[.ts]
--
include::flow-instructions/_upgrade-14-22-typescript.asciidoc[]
--

[.fusion]
--
include::fusion-instructions/_upgrade-14-22.asciidoc[]
--
```

Hide/reveal content based on the configuration by adding classes on the `<body>` element:

```html
<body class="v14-15 v15-16 v16-17 v17-18 v18-19 v19-20 v20-21 v21-22 v14-22 flow fusion spring-boot">

<style>
body:not(.v14-22) .v14-22,
body:not(.flow) .flow {
  display: none !important;
}
</style>
```

That requires generating those selectors when the upgrade tool is constructed, but that’s straightforward.

Author: `hidden `and `upgrade-instructions` classnames are now removed. Open blocks (`--`) are also now used. 

However, I didn't quite get how things would work by marking the version range only on the sections. If I have something like the following:
```
[.v14-15]
=== Upgrade Steps | 14 → 15

[.flow]
--
include::flow-instructions/_upgrade-14-15.asciidoc[]
--

[.v15-16]
=== Upgrade Steps | 15 → 16

[.flow]
--
include::flow-instructions/_upgrade-15-16.asciidoc[]
--
```

How can I show only the 14->15 flow instructions (without showing the 15->16 ones)?

Reviewer: Note, that I just played this in my head, I didn't actually try it, so there might be some case I didn't consider :)

But if the `.v15-16` section is hidden by default, then everything within it is going to be hidden. The content between the H3's (`=== ...`) is nested in the DOM within the respective H3 sections (`.sect2` class in the DOM). So even if you set the `.flow` class to be visible, the one inside the v15-16 section is not going to be visible.

Author: Ok, got you! Yes, you are right. It could be done how you are describing it. I just pushed a new commit that does just that.

---
title: Upgrading
order: 70
---

= Upgrading Guide
:!toc:

pass:[<!-- Vale Vaadin.Versions = NO -->]

Instructions for upgrading to the latest Vaadin version. To run applications or components developed with Vaadin 7 or 8 inside an application written using the latest version, see <<{articles}/tools/mpr/overview#,Multiplatform Runtime>>.

pass:[<!-- Vale Vaadin.Versions = YES -->]

[source,typescript, role="render-only embedded"]
----
include::{root}/frontend/demo/upgrade-tool/upgrade-tool.ts[render, persistent]
----

////
Upgrading instructions are all loaded to the page as hidden by default.

> The upgrade-tool manipulates the visibility of the instructions via their classnames depending of the user's selected options.

Reviewer 2: ```suggestion
The upgrade tool manipulates the visibility of the instructions via their classnames depending on the user's selected options.
```

This is the description of the classnames:

.all:
    Upgrading instructions that should appear regardless of the user's selected configurations.

.v{from-version}-{to-version}:
    For example, v14-17, will be used by the upgrade tool to show the marked instructions if the user selects to see the instructions for the v14 -> v17 path.

>     If there is no upgrading instructions directly for the v14 -> v17 path, the app will then display all the intermediary upgrade paths (i.e. v14->v15, v15->v16, and v16-v17).

Reviewer 2: ```suggestion
    If there are no upgrading instructions directly for the v14 -> v17 path, the app will then display all the intermediate upgrade paths (i.e. v14->v15, v15->v16, and v16-v17).
```

include::{root}/frontend/demo/upgrade-tool/upgrade-tool.ts[render, persistent]
----

////
Upgrading instructions are all loaded to the page as hidden by default.
The upgrade-tool manipulates the visibility of the instructions via their classnames depending of the user's selected options.

This is the description of the classnames:

.hidden:
    All items should be given this classname for them to be hidden when the page is loaded (the upgrade tool will remove this classname conditional on user selection).

.upgrade-instructions:
    All items should be given this classname as well. This classname will enable the upgrade tool to re-hide all the elements as the user selection changes.

.instructions-all:
    Upgrading instructions that should appear regardless of the user's selected configurations.

> .flow-instructions-all:

Reviewer: Having `-instructions` in all the selectors makes it quite verbose to write/author. Could we remove that, and the `-all` suffix as well, and go simply with `.flow`, `.fusion`, `.spring`, etc?

For version ranges, that would become `.flow-14-23`, etc.

Author: `-instructions` and `-all` suffixes are now removed.

.spring-instructions-{from-version}-{to-version}:
    Same as flow-instructions-{from-version}-{to-version} for any Spring project (regardless of whether it's flow or fusion based).

.typescript-instructions-{from-version}-{to-version}:
    Same as flow-instructions-{from-version}-{to-version} for any Typescript project (regardless of whether it's flow or fusion based).

.flow-instructions-spring-{from-version}-{to-version}:
    Same as flow-instructions-{from-version}-{to-version}, but only displays the content if the project is a Spring-based one.

.flow-instructions-typescript-{from-version}-{to-version}:
    Same as flow-instructions-{from-version}-{to-version}, but only displays the content if the project is a TypeScript-based one.

.fusion-instructions-spring-{from-version}-{to-version}:
    Same as fusion-instructions-{from-version}-{to-version}, but only displays the content if the project is a Spring-based one.

.fusion-instructions-typescript-{from-version}-{to-version}:
    Same as fusion-instructions-{from-version}-{to-version}, but only displays the content if the project is a TypeScript-based one.

////

[.hidden.upgrade-instructions.instructions-all]

> === Before you start

Reviewer: ```suggestion
=== Before You Start
```
